### PR TITLE
COOPR 542: Add ability to configure UI to use TLS

### PR DESCRIPTION
**Pull request by kpovetkin**
SSL/TLS implemented for coopr-ui.

### CODE REVIEW

**Comment by gaarf** [comment]
diff is hard to read because of numerous whitespace changes. Any way we can set our respective text editors to not change existing whitespace setttings?

File: coopr-ngui/server.js [hunk 1]
```diff
@@ -5,33 +5,49 @@
...
     express = require('express'),
     finalhandler = require('finalhandler'),
     serveFavicon = require('serve-favicon'),
     corsAnywhere = require('cors-anywhere'),
 
+    COOPR_SSL = ('true' == process.env.COOPR_SSL) || false;
     COOPR_UI_PORT = parseInt(process.env.COOPR_UI_PORT || 8080, 10),
+    COOPR_UI_SSL_PORT = parseInt(process.env.COOPR_UI_SSL_PORT || 8443, 10),
+    COOPR_UI_KEY_FILE = process.env.COOPR_UI_KEY_FILE || 'cert/server.key',
+    COOPR_UI_CERT_FILE = process.env.COOPR_UI_CERT_FILE || 'cert/server.crt',
```
Comment (gaarf) [inline coopr-ngui/server.js]: I thought we were going to use a .json file for config instead of env vars? cc @wolf31o2

File: coopr-ngui/server.js [hunk 2]
```diff
@@ -75,26 +97,37 @@ app.use('/assets', [
...
-  httpIndexLogger,
-  function (req, res) {
-    res.sendFile(__dirname + '/dist/index.html');
-  }
+    httpIndexLogger,
+    function (req, res) {
+        if (COOPR_SSL && sslStarted && !req.secure) {
+            res.redirect(['https://', req.hostname,
+                ':', COOPR_UI_SSL_PORT, req.originalUrl
+            ].join(''));
```
Comment (gaarf) [inline coopr-ngui/server.js]: offline conversation with @wolf31o2 implied we would not be doing both HTTP and HTTPS at the same time, but rather just one or the other.

**Comment by kpovetkin** [comment]
It's done according to CDAP project rules. Unfortunately I'm not able to get identation back. In the future we would not touch current identation.

File: coopr-ngui/server.js (hunk 1 above)
Comment (kpovetkin) [inline coopr-ngui/server.js]: We have no such requirements during development. It would be great if for this task there is a jira task.

File: coopr-ngui/server.js (hunk 2 above)
Comment (kpovetkin) [inline coopr-ngui/server.js]: In case https is enabled, https would be the only protocol user would work with.
It's a good practice to redirect user from http page to https page if SSL is enabled.

File: coopr-ngui/bin/ngui.sh [hunk 3]
```diff
@@ -15,10 +15,13 @@
...
-export COOPR_LOG_DIR=${COOPR_LOG_DIR:-/var/log/coopr}
-export COOPR_UI_PORT=${COOPR_UI_PORT:-8100}
-export COOPR_HOME=${COOPR_HOME:-/opt/coopr}
+COOPR_SERVER_URI=${COOPR_SERVER_URI:-http://localhost:55054}
+COOPR_LOG_DIR=${COOPR_LOG_DIR:-/var/log/coopr}
+COOPR_UI_PORT=${COOPR_UI_PORT:-8100}
+COOPR_UI_SSL_PORT=${COOPR_UI_SSL_PORT:-8443}
+COOPR_UI_KEY_FILE=${COOPR_UI_KEY_FILE:-'cert/server.key'}
+COOPR_UI_CERT_FILE=${COOPR_UI_CERT_FILE:-'cert/server.crt'}
+COOPR_HOME=${COOPR_HOME:-/opt/coopr} ; export COOPR_HOME
```
Comment (wolf31o2) [inline coopr-ngui/bin/ngui.sh]: Do not revert my changes to this script. Fix your conflict here by conforming to the current way the script works, please. You need to export these variables.

Also, use absolute PATH for COOPR_UI_KEY_FILE and COOPR_UI_CERT_FILE, do not use quotes within shell expansion.

``` bash
export COOPR_UI_KEY_FILE=${COOPR_UI_KEY_FILE:-${COOPR_HOME}/ngui/certs/server.key}
```

File: coopr-ngui/bin/ngui.sh [hunk 4]
```diff
@@ -48,6 +51,11 @@ start ( ) {
 
   echo "Starting Coopr NGUI ..."
 
+  export COOPR_UI_PORT
+  export COOPR_UI_SSL_PORT
+  export COOPR_UI_KEY_FILE
+  export COOPR_UI_CERT_FILE
+
```
Comment (wolf31o2) [inline coopr-ngui/bin/ngui.sh]: Remove these. They should be exported above.

File: coopr-ngui/server.js (hunk 1 above)
Comment (wolf31o2) [inline coopr-ngui/server.js]: Let's get this tested/merged, then change all of the variables to a config file, all at once.

File: coopr-ngui/server.js (hunk 2 above)
Comment (wolf31o2) [inline coopr-ngui/server.js]: I can see both ways. The intent was to have the UI only listening on either HTTP or HTTPS, but I can see the advantage to both. I specifically requested the Coopr Cloud UI do this redirection, so there's definitely use cases for it. Maybe have some option to control whether we do this redirect?

File: coopr-provisioner/master/conf/provisioner-site.xml [hunk 5]
```diff
@@ -18,9 +18,13 @@
...
-  -->
-
+    <!--
+      Your provisioner site level configuration goes here
+    -->
+    <property>
+        <name>provisioner.server.uri</name>
+        <value>https://localhost:55054</value>
+        <description>URI of server to connect to</description>
+    </property>
```
Comment (wolf31o2) [inline coopr-provisioner/master/conf/provisioner-site.xml]: Remove this change. This setting is the default.

File: coopr-standalone/bin/coopr.sh [hunk 6]
```diff
@@ -133,21 +133,71 @@ elif [ ${RUBY_VERSION_MAJOR} -eq 1 ] && [ ${RUBY_VERSION_MINOR} -lt 9 ]; then
   die "Ruby version is not supported! The minimum version supported is v1.9.0p0"
 fi
 
+# $1 - Property name to read
+# $2 - Config file name to read from
+# $3 - variable for return value
+#
+# Usage: read_property property_name /path/to/config/file variable_to_store_value
+#
+function read_property {
```
Comment (wolf31o2) [inline coopr-standalone/bin/coopr.sh]: The "function" label isn't necessary and causes issues on some shells.

File: coopr-standalone/bin/coopr.sh [hunk 7]
```diff
@@ -133,21 +133,71 @@ elif [ ${RUBY_VERSION_MAJOR} -eq 1 ] && [ ${RUBY_VERSION_MINOR} -lt 9 ]; then
...
+    local __result_var=$3
```
Comment (wolf31o2) [inline coopr-standalone/bin/coopr.sh]: Don't use "local" as it isn't portable.

File: coopr-standalone/bin/coopr.sh [hunk 8]
```diff
@@ -133,21 +133,71 @@ elif [ ${RUBY_VERSION_MAJOR} -eq 1 ] && [ ${RUBY_VERSION_MINOR} -lt 9 ]; then
...
+    local property_re='(?<=<name>'$1'</name>)[\s\S]+?(?=</property>)'
+    local property_value_re='(?<=<value>)[\s\S]+?(?=</value>)'
+
+    eval $__result_var=`grep -Pzoe $property_re $2 | grep -Pzoe $property_value_re`
+}
```
Comment (wolf31o2) [inline coopr-standalone/bin/coopr.sh:148]: This function should return the value, directly. Do not assign it to a variable.

File: coopr-standalone/bin/coopr.sh [hunk 9]
```diff
@@ -133,21 +133,71 @@ elif [ ${RUBY_VERSION_MAJOR} -eq 1 ] && [ ${RUBY_VERSION_MINOR} -lt 9 ]; then
...
+
 # Setup coopr configuration
 COOPR_PROTOCOL=http
-line=`awk '/server.ssl.enable/{print NR; exit}' ${COOPR_SERVER_CONF}/coopr-site.xml`
-let "line++"
-ssl=`sed -n "${line}p" ${COOPR_SERVER_CONF}/coopr-site.xml | awk -F"<|>" '{print $3}'`
-if [ ! -z ${ssl} ] && [ ${ssl} = "true" ]; then
-  COOPR_PROTOCOL=https
+read_property server.ssl.enabled ${COOPR_SERVER_CONF}coopr-site.xml COOPR_SSL
+export COOPR_SSL
```
Comment (wolf31o2) [inline coopr-standalone/bin/coopr.sh]: These should be replaced with:

``` bash
export COOPR_SSL=`read_property server.ssl.enabled ${COOPR_SERVER_CONF}/coopr-site.xml`
```

File: coopr-standalone/bin/coopr.sh [hunk 10]
```diff
@@ -133,21 +133,71 @@ elif [ ${RUBY_VERSION_MAJOR} -eq 1 ] && [ ${RUBY_VERSION_MINOR} -lt 9 ]; then
...
+if [ ! -z $COOPR_SSL ] && [ $COOPR_SSL = "true" ]; then
```
Comment (wolf31o2) [inline coopr-standalone/bin/coopr.sh]: Don't use negative matching. Use `-n` instead of the inverse of `-z` and always wrap your variables in braces.

File: coopr-standalone/bin/coopr.sh [hunk 11]
```diff
@@ -133,21 +133,71 @@ elif [ ${RUBY_VERSION_MAJOR} -eq 1 ] && [ ${RUBY_VERSION_MINOR} -lt 9 ]; then
...
+    COOPR_PROTOCOL=https
+
+    read_property server.nodejs.ssl.path ${COOPR_SERVER_CONF}coopr-security.xml COOPR_NODEJS_SSL_PATH
```
Comment (wolf31o2) [inline coopr-standalone/bin/coopr.sh]: Nothing guarantees that this property is not set in `coopr-site.xml` instead.

Also, why are we not using the same variables as the UI, itself? We should be using COOPR_UI_KEY_FILE and COOPR_UI_CERT_FILE variables.

File: coopr-standalone/bin/coopr.sh [hunk 12]
```diff
@@ -133,21 +133,71 @@ elif [ ${RUBY_VERSION_MAJOR} -eq 1 ] && [ ${RUBY_VERSION_MINOR} -lt 9 ]; then
...
 
+read_property server.ssl.trust.cert.path ${COOPR_SERVER_CONF}coopr-security.xml TRUST_CERT_PATH
+export TRUST_CERT_PATH
+
+read_property server.ssl.trust.cert.password ${COOPR_SERVER_CONF}coopr-security.xml TRUST_CERT_PASSWORD
+export TRUST_CERT_PASSWORD
+
+if [ ! -z TRUST_CERT_PATH ] && [ ! -z TRUST_CERT_PASsWORD ]; then
+  export CERT_PARAMETER="--cert ${TRUST_CERT_PATH}:${TRUST_CERT_PASSWORD}"
+fi
```
Comment (wolf31o2) [inline coopr-standalone/bin/coopr.sh:173]: Negative conditional. Typo in variable name....

...and curl does _NOT_ support Java Keystores. This does not work.

File: coopr-standalone/bin/coopr.sh [hunk 13]
```diff
@@ -133,21 +133,71 @@ elif [ ${RUBY_VERSION_MAJOR} -eq 1 ] && [ ${RUBY_VERSION_MINOR} -lt 9 ]; then
...
 if [ ${COOPR_PROTOCOL} = "https" ]; then
   export CURL_PARAMETER="--insecure"
   export COOPR_REJECT_UNAUTH=false
 fi
 
+read_property server.ssl.trust.keystore.path ${COOPR_SERVER_CONF}coopr-security.xml keystore_path
+read_property server.ssl.trust.keystore.password ${COOPR_SERVER_CONF}coopr-security.xml keystore_password
+
+COOPR_NODE_TLS_ENABLED="false"
+if [ ! -z keystore_path ] && [ ! -z keystore_password ]; then
```
Comment (wolf31o2) [inline coopr-standalone/bin/coopr.sh]: More negative conditionals. Please, replace these with positive conditionals.

File: coopr-standalone/server/conf/coopr-security.xml [hunk 14]
```diff
@@ -2,8 +2,67 @@
...
+        <name>server.ssl.keystore.password</name>
+        <value>secret</value>
+    </property>
+    <property>
+        <name>server.ssl.cert.password</name>
+        <value>secret</value>
+    </property>
+    <property>
+        <name>server.ssl.trust.keystore.path</name>
+        <value>/home/mak/provisioner.jks</value>
```
Comment (wolf31o2) [inline coopr-standalone/server/conf/coopr-security.xml]: Invalid PATH.

File: coopr-standalone/server/conf/coopr-security.xml [hunk 15]
```diff
@@ -2,8 +2,67 @@
...
+    </property>
+    <property>
+        <name>server.ssl.trust.keystore.password</name>
+        <value>123456</value>
+    </property>
+    <property>
+        <name>server.ssl.trust.cert.path</name>
+        <value>/home/mak/provisioner.pem</value>
```
Comment (wolf31o2) [inline coopr-standalone/server/conf/coopr-security.xml]: Invalid PATH.

File: coopr-standalone/server/conf/coopr-security.xml [hunk 16]
```diff
@@ -2,8 +2,67 @@
...
+    </property>
+    <property>
+        <name>server.ssl.trust.cert.password</name>
+        <value>123456</value>
+    </property>
+    <property>
+        <name>server.nodejs.tls.cert.path</name>
+        <value></value>
+    </property>
```
Comment (wolf31o2) [inline coopr-standalone/server/conf/coopr-security.xml]: Why is there UI configuration in the server's config file?

File: coopr-ui/bin/ui.sh [hunk 17]
```diff
@@ -50,6 +51,7 @@ start ( ) {
   echo "Starting UI ..."
   nohup nice -1 ${COOPR_NODE} ${UI_PATH}/server.js --env=${ENVIRONMENT} \
     --cooprhost=${COOPR_SERVER_URI} --port=${COOPR_UI_PORT} \
+    --ssl=${COOPR_SSL} --sslport=${COOPR_UI_SSL_PORT} \
```
Comment (wolf31o2) [inline coopr-ui/bin/ui.sh:55]: You need to default COOPR_SSL. You cannot assume that it will be set to the correct value, especially on installs versus standalone.

File: coopr-ui/server.js [hunk 18]
```diff
@@ -25,13 +25,17 @@ var express = require('express'),
...
+    nock = require('nock'),
+    http = require('http'),
+    https = require('https');
 
 /**
  * Set environment vars.
  */
 var env = argv.env || 'production';
 var PORT = argv.port || 8100;
+var SSL = ('true' == argv.ssl);
```
Comment (wolf31o2) [inline coopr-ui/server.js:37]: Do not change default. We should not be SSL by default.

File: coopr-standalone/bin/coopr.sh (hunk 11 above)
Comment (kpovetkin) [inline coopr-standalone/bin/coopr.sh]: 1. If some properties are not set something would not work and there would logs or unexpected behaviour.
2. COOPR_UI_KEY_FILE and COOPR_UI_CERT_FILE are used to crypt http server on node, while tls params are used to make tls connection working between nodejs and coopr server.

File: coopr-standalone/bin/coopr.sh (hunk 12 above)
Comment (kpovetkin) [inline coopr-standalone/bin/coopr.sh:173]: Unfortunately it was not done by me. I'll ask guys to review this code.

File: coopr-standalone/server/conf/coopr-security.xml (hunk 14 above)
Comment (kpovetkin) [inline coopr-standalone/server/conf/coopr-security.xml]: Unfortunately it was not done by me. I'll ask guys to review this code.

File: coopr-ui/bin/ui.sh (hunk 17 above)
Comment (kpovetkin) [inline coopr-ui/bin/ui.sh:55]: Added default value.

File: coopr-ngui/server.js (hunk 1 above)
Comment (wolf31o2) [inline coopr-ngui/server.js]: These should use an absolute PATH, not relative PATH.

File: coopr-ngui/server.js [hunk 19]
```diff
@@ -5,33 +5,49 @@
...
+    fs = require('fs'),
     morgan = require('morgan'),
+    http = require('http'),
+    https = require('https'),
     express = require('express'),
     finalhandler = require('finalhandler'),
     serveFavicon = require('serve-favicon'),
     corsAnywhere = require('cors-anywhere'),
 
+    COOPR_SSL = ('true' == process.env.COOPR_SSL) || false;
```
Comment (wolf31o2) [inline coopr-ngui/server.js]: This should default to false. Currently, it will be true if `COOPR_SSL=false` is exported.

File: coopr-provisioner/master/conf/provisioner-site.xml [hunk 20]
```diff
@@ -18,9 +18,8 @@
 
 -->
 <configuration>
-  <!--
-    Your provisioner site level configuration goes here
-  -->
-
+    <!--
+      Your provisioner site level configuration goes here
+    -->
```
Comment (wolf31o2) [inline coopr-provisioner/master/conf/provisioner-site.xml]: This isn't necessary. Revert this file to the original.

File: coopr-standalone/bin/coopr.sh [hunk 21]
```diff
@@ -133,21 +133,73 @@ elif [ ${RUBY_VERSION_MAJOR} -eq 1 ] && [ ${RUBY_VERSION_MINOR} -lt 9 ]; then
...
+}
+
 # Setup coopr configuration
 COOPR_PROTOCOL=http
-line=`awk '/server.ssl.enable/{print NR; exit}' ${COOPR_SERVER_CONF}/coopr-site.xml`
-let "line++"
-ssl=`sed -n "${line}p" ${COOPR_SERVER_CONF}/coopr-site.xml | awk -F"<|>" '{print $3}'`
-if [ ! -z ${ssl} ] && [ ${ssl} = "true" ]; then
-  COOPR_PROTOCOL=https
+COOPR_SSL=`read_property server.ssl.enabled ${COOPR_SERVER_CONF}coopr-site.xml`
```
Comment (wolf31o2) [inline coopr-standalone/bin/coopr.sh]: All instances of this are missing a `/` between `${COOPR_SERVER_CONF}` and `coopr-site.xml` which needs to be added.

File: coopr-standalone/bin/coopr.sh [hunk 22]
```diff
@@ -133,21 +133,73 @@ elif [ ${RUBY_VERSION_MAJOR} -eq 1 ] && [ ${RUBY_VERSION_MINOR} -lt 9 ]; then
...
+if [ -z $COOPR_SSL ]; then
+    COOPR_SSL="false"
 fi
```
Comment (wolf31o2) [inline coopr-standalone/bin/coopr.sh]: Convert this to a single line... there's no need for `if` if you have no else and it isn't a compound statement.

``` bash
[ -z ${COOPR_SSL} ] && COOPR_SSL=false
```

File: coopr-standalone/bin/coopr.sh [hunk 23]
```diff
@@ -133,21 +133,73 @@ elif [ ${RUBY_VERSION_MAJOR} -eq 1 ] && [ ${RUBY_VERSION_MINOR} -lt 9 ]; then
...
+# $2 - Config file name to read from
+# $3 - variable for return value
+#
+# Usage: read_property property_name /path/to/config/file variable_to_store_value
+#
+read_property () {
+    property_re='(?<=<name>'$1'</name>)[\s\S]+?(?=</property>)'
+    property_value_re='(?<=<value>)[\s\S]+?(?=</value>)'
+
+    echo `grep -Pzoe $property_re $2 | grep -Pzoe $property_value_re`
```
Comment (wolf31o2) [inline coopr-standalone/bin/coopr.sh]: 2 character indentation, not 4...

File: coopr-standalone/bin/coopr.sh [hunk 24]
```diff
@@ -133,21 +133,73 @@ elif [ ${RUBY_VERSION_MAJOR} -eq 1 ] && [ ${RUBY_VERSION_MINOR} -lt 9 ]; then
   die "Ruby version is not supported! The minimum version supported is v1.9.0p0"
 fi
 
+# $1 - Property name to read
+# $2 - Config file name to read from
+# $3 - variable for return value
+#
+# Usage: read_property property_name /path/to/config/file variable_to_store_value
```
Comment (wolf31o2) [inline coopr-standalone/bin/coopr.sh]: Update these comments to reflect the current version of the function.

File: coopr-standalone/bin/coopr.sh [hunk 25]
```diff
@@ -133,21 +133,73 @@ elif [ ${RUBY_VERSION_MAJOR} -eq 1 ] && [ ${RUBY_VERSION_MINOR} -lt 9 ]; then
...
+    COOPR_SSL="false"
 fi
+export COOPR_SSL
+if [ $COOPR_SSL = "true" ]; then
+    COOPR_PROTOCOL=https
+
+    COOPR_NODEJS_SSL_PATH=`read_property server.nodejs.ssl.path ${COOPR_SERVER_CONF}coopr-security.xml`
+    COOPR_NODEJS_SSL_KEY=`read_property server.nodejs.ssl.key ${COOPR_SERVER_CONF}coopr-security.xml`
+    export COOPR_NODEJS_SSL_KEY=$COOPR_NODEJS_SSL_PATH"/"$COOPR_NODEJS_SSL_KEY
+    COOPR_NODEJS_SSL_CRT=`read_property server.nodejs.ssl.crt ${COOPR_SERVER_CONF}coopr-security.xml`
```
Comment (wolf31o2) [inline coopr-standalone/bin/coopr.sh]: Use a different interim variable for COOPR_NODEJS_SSL_KEY and COOPR_NODEJS_SSL_CRT rather than overriding them...

or...

Have the node scripts use both COOPR_NODEJS_SSL_PATH and the KEY/CRT variables versus requiring the KEY/CRT variables to include the full PATH and filename.

Comment (wolf31o2) [inline coopr-standalone/bin/coopr.sh]: Also, missing a `/` like above...

File: coopr-standalone/bin/coopr.sh [hunk 26]
```diff
@@ -133,21 +133,73 @@ elif [ ${RUBY_VERSION_MAJOR} -eq 1 ] && [ ${RUBY_VERSION_MINOR} -lt 9 ]; then
...
+    export COOPR_NODEJS_SSL_CRT=$COOPR_NODEJS_SSL_PATH"/"$COOPR_NODEJS_SSL_CRT
+fi
+
 export COOPR_SERVER_URI=${COOPR_PROTOCOL}://localhost:55054
 
+read_property server.ssl.trust.cert.path ${COOPR_SERVER_CONF}coopr-security.xml TRUST_CERT_PATH
+export TRUST_CERT_PATH
+
+read_property server.ssl.trust.cert.password ${COOPR_SERVER_CONF}coopr-security.xml TRUST_CERT_PASSWORD
+export TRUST_CERT_PASSWORD
```
Comment (wolf31o2) [inline coopr-standalone/bin/coopr.sh]: These no longer work with the changes to `read_property` above...

``` bash
export TRUST_CERT_PATH=`read_property server.ssl.trust.cert.path ${COOPR_SERVER_CONF}/coopr-security.xml`
```

Please, note the `/` after the `${COOPR_SERVER_CONF}` which is required.

File: coopr-standalone/bin/coopr.sh [hunk 27]
```diff
@@ -133,21 +133,73 @@ elif [ ${RUBY_VERSION_MAJOR} -eq 1 ] && [ ${RUBY_VERSION_MINOR} -lt 9 ]; then
...
+
+COOPR_NODE_TLS_ENABLED="false"
+if [ ! -z keystore_path ] && [ ! -z keystore_password ]; then
+    COOPR_NODE_TLS_ENABLED="true"
+
+    read_property server.nodejs.tls.cert.path ${COOPR_SERVER_CONF}coopr-security.xml nodejs_tls_path
+    read_property server.nodejs.tls.key ${COOPR_SERVER_CONF}coopr-security.xml nodejs_tls_key
+    read_property server.nodejs.tls.crt ${COOPR_SERVER_CONF}coopr-security.xml nodejs_tls_crt
+    read_property server.nodejs.tls.ca ${COOPR_SERVER_CONF}coopr-security.xml nodejs_tls_ca
+    read_property server.nodejs.tls.password ${COOPR_SERVER_CONF}coopr-security.xml COOPR_NODE_TLS_PASSWORD
```
Comment (wolf31o2) [inline coopr-standalone/bin/coopr.sh]: None of these will work. Update to use new `read_property` parameters and assign it to a variable.

File: coopr-standalone/server/conf/coopr-security.xml [hunk 28]
```diff
@@ -2,8 +2,7 @@
 <?xml-stylesheet type="text/xsl" href="configuration.xsl"?>
 <!-- Copyright © 2012-2014 Cask Data, Inc. All Rights Reserved.  -->
 <configuration>
-  <!--
-    Your security level configuration goes here
-  -->
-
+    <!--
+      Your security level configuration goes here
+    -->
```
Comment (wolf31o2) [inline coopr-standalone/server/conf/coopr-security.xml]: Revert this unnecessary change.

File: coopr-standalone/server/conf/coopr-site.xml [hunk 29]
```diff
@@ -2,12 +2,12 @@
...
-    <value>COOPR_DATA_DIR/plugins/resources</value>
-  </property>
+    <property>
+        <name>server.local.data.dir</name>
+        <value>COOPR_DATA_DIR</value>
+    </property>
+    <property>
+        <name>server.plugin.store.localfilestore.data.dir</name>
+        <value>COOPR_DATA_DIR/plugins/resources</value>
+    </property>
```
Comment (wolf31o2) [inline coopr-standalone/server/conf/coopr-site.xml]: Revert changes to this file.

File: coopr-ui/bin/ui.sh (hunk 17 above)
Comment (wolf31o2) [inline coopr-ui/bin/ui.sh:55]: Where?

I was expecting to see something like:

``` bash
export COOPR_SSL=${COOPR_SSL:-false}
```

**Comment by wolf31o2** [comment]
Also, this needs to be updated with latest changes from develop. It doesn't merge cleanly.

File: coopr-ngui/server.js (hunk 19 above)
Comment (kpovetkin) [inline coopr-ngui/server.js]: It would not. NodeJS evaluates environment variables into string values, so this check is correct. Moreover this check would work correct in case when environment variable is not set at all.

The only thing should be corrected there is '|| false'.

http://codepen.io/anon/pen/gCKny?editors=101

**Comment by kpovetkin** [comment]
Merge confilcts are resolved

File: coopr-ngui/server.js [hunk 30]
```diff
@@ -75,26 +98,33 @@ app.use('/assets', [
...
 ]);
 
-app.listen(COOPR_UI_PORT, '0.0.0.0', function () {
-  console.info(httpLabel+' listening on port %s', COOPR_UI_PORT);
-});
-
+if (!COOPR_SSL) {
+    httpServer.listen(COOPR_UI_PORT, null, null, function () {
+        console.info(httpLabel + ' listening on port %s', COOPR_UI_PORT);
+    });
```
Comment (cdap-apps) [inline coopr-ngui/server.js:124]:
```
COOPR_EFFECTIVE_PORT=COOPR_UI_SSL_PORT
if(!COOPR_SSL) {
  COOPR_EFFECTIVE_PORT = COOPR_UI_PORT
}

httpServer.listen(COOPR_EFFECTIVE_PORT, null, null, function() {
   console.info(httpLabel + ' listening on port %s', COOPR_EFFECTIVE_PORT);
});

```

File: coopr-ngui/server.js [hunk 31]
```diff
@@ -40,27 +57,33 @@ console.log(color.hilite(pkg.name) + ' v' + pkg.version + ' starting up...');
  * HTTP server
  */
 
-var app = express();
+var app = express(),
+    sslCredentials = {
+        key: fs.readFileSync(COOPR_UI_KEY_FILE, 'utf-8'),
+        cert: fs.readFileSync(COOPR_UI_CERT_FILE, 'utf-8')
+    },
```
Comment (gaarf) [inline coopr-ngui/server.js]: I think if I'm running with COOPR_SSL undefined or set to `false` (eg locally on developer's laptop), then I should be able to start the server even if the key and/or certificate files are missing.

Comment (kpovetkin) [inline coopr-ngui/server.js]: corrected

File: coopr-ngui/server.js (hunk 30 above)
Comment (kpovetkin) [inline coopr-ngui/server.js:124]: It would not work as long as httpServer and httpsServer are instances of different types.

Comment (gaarf) [inline coopr-ngui/server.js:124]: what @cdap-apps suggests makes sense to me, not sure what to issue would be.

**Comment by gaarf** [comment]
Changes to `coopr-ngui` LGTM. Have not looked at `coopr-ui` as I expect we will remove the old UI soon.

**Comment by wolf31o2** [comment]
@gaarf the shell is all good. If you're good with the UI code, can we merge?

File: coopr-ui/server.js [hunk 32]
```diff
@@ -1507,25 +1539,47 @@ site.app.get('/*', function(req, res) {
...
+    site.httpServer = http.createServer(site.app);
+    site.httpServer.listen(site.PORT, null, null, function () {
+        site.logger.info('Server started on port ', site.PORT);
+    });
+}
+
+if (SSL) {
+    try {
+        var COOPR_UI_KEY_FILE = process.env['COOPR_NODEJS_SSL_KEY'],
+            COOPR_UI_CERT_FILE = process.env['COOPR_NODEJS_SSL_CRT'],
```
Comment (gaarf) [inline coopr-ui/server.js:1569]: in the `ngui`, the env var is `process.env.COOPR_UI_KEY_FILE` ... why is a different one used here?

File: coopr-ngui/server.js (hunk 30 above)
Comment (kpovetkin) [inline coopr-ngui/server.js:124]: http serves http request only. https server - http/ssl respectively. http, https servers have different constructors, so it's not possible to implement @cdap-apps asked for.

File: coopr-ui/server.js (hunk 32 above)
Comment (kpovetkin) [inline coopr-ui/server.js:1569]: ngui has no complete ssl/tls support, because of task requirements.
ui implementation uses one key group to serve users over https, the second one to be able to communicate with tls encripted coopr server.

File: coopr-ngui/server.js (hunk 30 above)
Comment (gaarf) [inline coopr-ngui/server.js:124]:
```
var port = COOPR_UI_PORT,
    httpServer = http.createServer(app);
if(COOPR_SSL) {
   httpServer = https.createServer(app);
   port = COOPR_UI_SSL_PORT;
}
httpServer.listen(port, null, null, function() {
   console.info(httpLabel + ' listening on port %s using %s protocol', port, 
      COOPR_SSL ? "https" : "http");
});
```

duck typing FTW.

**Comment by gaarf** [comment]
all righty :+1:
